handlers/freshchat/freshchat.go

```diff
@@ -0,0 +1,275 @@
+		text := msg.Text()
+		var msgtext = new(MessageParts)
+		msgtext.Text = &Text{Content: text}
+		payload.Messages[0].MessageParts = append(payload.Messages[0].MessageParts, *msgtext)
+	}
+
+	if len(msg.Attachments()) > 0 {
+		mediaURL := msg.Attachments()[0]
+		var msgimage = new(MessageParts)
+		msgimage.Image = &Image{URL: mediaURL}
+		payload.Messages[0].MessageParts = append(payload.Messages[0].MessageParts, *msgimage)
+	}
+
+	jsonBody, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(jsonBody))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	var bearer = "Bearer " + authToken
+	req.Header.Set("Authorization", bearer)
+
+	rr, err := utils.MakeHTTPRequest(req)
+
+	// record our status and log
+	log := courier.NewChannelLogFromRR("Message Sent", msg.Channel(), msg.ID(), rr).WithError("Message Send Error", err)
+	status.AddLog(log)
+	if err != nil {
+		return status, err
+	}
+	status.SetStatus(courier.MsgWired)
+
+	return status, nil
+}
+
+func (h *handler) validateSignature(c courier.Channel, r *http.Request) error {
+	if !h.validateSignatures {
+		return nil
+	}
+	var rsaPubKey = []byte(c.StringConfigForKey(courier.ConfigPassword, ""))
+
+	actual := r.Header.Get(signatureHeader)
+	if actual == "" {
+		return fmt.Errorf("missing request signature")
+	}
+	buf, _ := ioutil.ReadAll(r.Body)
+	rdr1 := ioutil.NopCloser(bytes.NewBuffer(buf))
+	rdr2 := ioutil.NopCloser(bytes.NewBuffer(buf))
+	token, err := ioutil.ReadAll(rdr1)
+	if err != nil {
+		return fmt.Errorf("unable to read Body, %s", err.Error())
+	}
+	r.Body = rdr2
+
+	var b64Sig = []byte(actual)
+	block, _ := pem.Decode(rsaPubKey)
+	if err != nil {
+		return fmt.Errorf("failed to decode public key, %s", err.Error())
+	}
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("failed to parse DER encoded public key, %s", err.Error())
+	}
+	hash := sha256.New()
+	if _, err := bytes.NewReader(token).WriteTo(hash); err != nil {
+		return fmt.Errorf("unable to hash signed token, %s", err.Error())
+	}
+	decodedSig, err := base64.StdEncoding.DecodeString(string(b64Sig))
+	if err != nil {
+		return fmt.Errorf("unable to decode base64 signature, %s", err.Error())
+	}
+
+	if err := rsa.VerifyPKCS1v15(pub.(*rsa.PublicKey), crypto.SHA256, hash.Sum(nil), decodedSig); err != nil {
+		return fmt.Errorf("unable to verify signature, %s", err.Error())
+	}
+
+	return nil
+}
+
+type messagePayload struct {
+	Messages  []Messages `json:"messages"`
+	Status    string     `json:"status,omitempty"`
+	ChannelID string     `json:"channel_id"`
+	Users     []Users    `json:"users"`
+}
+type Messages struct {
+	MessageParts []MessageParts `json:"message_parts"`
+	ActorID      string         `json:"actor_id"`
+	ActorType    string         `json:"actor_type"`
+}
+
+type Users struct {
+	ID string `json:"id"`
+}
+type moPayload struct {
```

Comment:
so the only reason I didn't was that I'm reusing some of the structs between send and receive, but I can definitely do that for the receive payload. For the sending payload I think it makes it clearer to avoid inner structs so they don't have to be redefined while building the the outer structs. Thoughts?

Reply:
Ya, I think that's a fair point if there are pieces that are reused.